perf run all hv_24x7 events

nasastry opened the pull request:
On any machine Power8 or Power9 dynamically get all HV_24x7
events and according to the event's syntax fill the details and
run them.

Signed-off-by: Nageswara R Sastry <rnsastry@linux.vnet.ibm.com>

nasastry commented:
Test result `Pass` job log:
```
JOB ID     : 534a6b879ff323c6daf444753bdff929c3440cbf
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-03-28T16.27-534a6b8/job.log
 (1/1) perf_24x7_all_events.py:hv_24x7_all_events.test_all_events:
PASS (4695.41 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 4695.79 s
JOB HTML   : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-03-28T16.27-534a6b8/results.html
```

Test result `Fail` job log

```
JOB ID     : 534a6b879ff323c6daf444753bdff929c3440cbf
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-03-28T16.27-534a6b8/job.log
 (1/1) perf_24x7_all_events.py:hv_24x7_all_events.test_all_events:
PASS (4695.41 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 4695.79 s
JOB HTML   : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-03-28T16.27-534a6b8/results.html
```

PraveenPenguin commented:
Depend on https://github.com/avocado-framework/avocado/pull/3081

nasastry commented:
https://github.com/avocado-framework/avocado/pull/3081 got merged. Please review this PR.

harish-24 commented on perf/perf_24x7_all_events.py:
```diff
@@ -0,0 +1,145 @@
+            self.cancel("This test is not supported on guest")
+
+        # Check if 24x7 is present
+        if os.path.exists("%s" % self.event_sysfs):
```
directly use as `os.path.exists(self.event_sysfs)`?

harish-24 commented on perf/perf_24x7_all_events.py:
```diff
@@ -0,0 +1,145 @@
+        output = process.run('numactl -H')
+        for line in output.stdout.split('\n'):
+            if 'available' in line:
+                self.chip = int(line.split('-')[1].split(')')[0].strip())
```
This will fail on machines without range. ex. `available: 2 nodes (0,8)`

nasastry replied:
@harish-24 In some of the machines I can see `available: 2 nodes (0-1)`. As you pointed out, it might be an issue with the non-continuous nodes. I will have a look. Thanks!!

harish-24 commented on perf/perf_24x7_all_events.py:
```diff
@@ -0,0 +1,145 @@
+
+    def tearDown(self):
+        if os.path.isfile(self.perf_list_file):
+            output = process.run("rm -f %s" % self.perf_list_file)
```
`output` is not required.

harish-24 commented on perf/perf_24x7_all_events.py:
```diff
@@ -0,0 +1,145 @@
+        if os.path.isfile(self.perf_list_file):
+            output = process.run("rm -f %s" % self.perf_list_file)
+        # Collect the dmesg
+        output = process.run("dmesg -T")
```
here too.

harish-24 commented:
@nasastry @narasimhan-v @PraveenPenguin What is your opinion on creating a perf utility in `avocado`? I see many test re-using the same code. It can return a file with the list of events.
```
        self.perf_list_file = tempfile.NamedTemporaryFile().name
        output = process.run("perf list")
        for line in output.stdout.split('\n'):
            if 'hv_24x7' in line:
                line = line.split(',')[0].split('/')[1]
                genio.append_one_line(self.perf_list_file, line)
```

nasastry commented:
> @nasastry @narasimhan-v @PraveenPenguin What is your opinion on creating a perf utility in `avocado`? I see many test re-using the same code. It can return a file with the list of events.
> 
> ```
>         self.perf_list_file = tempfile.NamedTemporaryFile().name
>         output = process.run("perf list")
>         for line in output.stdout.split('\n'):
>             if 'hv_24x7' in line:
>                 line = line.split(',')[0].split('/')[1]
>                 genio.append_one_line(self.perf_list_file, line)
> ```

@harish-24 that make sense. I can write one.

nasastry commented:
@harish-24 Implemented the changes please review.

On PowerVM machine where the HMC is not enabled to collect events:
```
JOB ID     : 34542593d3438fb177e400979aac719236b19e91
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-04-02T04.46-3454259/job.log
 (1/1) perf_24x7_all_events.py:hv_24x7_all_events.test_all_events:  CANCEL: Please enable LPAR to allow collecting the 24x7 counters info (1.34 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 1
JOB TIME   : 1.63 s
JOB HTML   : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-04-02T04.46-3454259/results.html
```

On PowerNV machine:
```
JOB ID     : 88def22466fd822b79833e20f80db2bed90bdcf5
JOB LOG    : /var/lib/libvirt/images/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-04-02T04.58-88def22/job.log
 (1/1) perf_24x7_all_events.py:hv_24x7_all_events.test_all_events: CANCEL: /sys/bus/event_source/devices/hv_24x7 doesn't exist.This feature is supported only on LPAR (1.02 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 1
JOB TIME   : 1.54 s
JOB HTML   : /var/lib/libvirt/images/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-04-02T04.58-88def22/results.html
```

On Power8 machine:
```
JOB ID     : 96cd3ae3078d21fd8bb14b59230c9e3684ede2d3
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-04-02T06.12-96cd3ae/job.log
 (1/1) perf_24x7_all_events.py:hv_24x7_all_events.test_all_events:  PASS (3401.17 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 3401.49 s
JOB HTML   : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-04-02T06.12-96cd3ae/results.html
```

On Power9 machine:
```
JOB ID     : c72dd700852feeee7241582449e69be9fb05b50e
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-04-02T15.43-c72dd70/job.log
 (1/1) perf_24x7_all_events.py:hv_24x7_all_events.test_all_events:  PASS (4695.25 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 4695.65 s
JOB HTML   : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-04-02T15.43-c72dd70/results.html
```

nasastry commented:
@harish-24 Added perf utility and pushed the test code.

perf utility PR is at https://github.com/avocado-framework/avocado/pull/3090

```
JOB ID     : 52aff0b9b1b33712b4c55dd19fd25052e376a8ad
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-04-04T13.57-52aff0b/job.log
 (1/1) perf_24x7_all_events.py:hv_24x7_all_events.test_all_events:  PASS (4697.25 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 4697.61 s
JOB HTML   : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-04-04T13.57-52aff0b/results.html
```

nasastry commented:
@PraveenPenguin Dependency patch https://github.com/avocado-framework/avocado/pull/3108 got merged. Please merge this also.

PraveenPenguin commented on perf/perf_24x7_all_events.py:
```diff
@@ -0,0 +1,135 @@
+        """
+        smm = SoftwareManager()
+        detected_distro = distro.detect()
+        if 'ppc' not in process.system_output("uname -p", ignore_status=True):
```
distro.arch can be used and even string also need be changed as ppc64 stands for BE version 

like 
```
if 'ppc' not in detected_distro.arch:
    self.cancel("Processor is not PowerPC")
```

nasastry replied:
done

PraveenPenguin commented on perf/perf_24x7_all_events.py:
```diff
@@ -0,0 +1,135 @@
+        # Check if this is a guest
+        # 24x7 is not suported on guest
+        cpu_output = process.run("cat /proc/cpuinfo")
+        if "emulated by" in cpu_output.stdout:
```
avocado lib can be used here as well 

`if "emulated by" in  _get_cpu_info():`

nasastry replied:
done

PraveenPenguin commented on perf/perf_24x7_all_events.py:
```diff
@@ -0,0 +1,135 @@
+            self.log.info('hv_24x7 present')
+        else:
+            self.cancel("%s doesn't exist.This feature is supported"
+                        " only on LPAR" % self.event_sysfs)
```
can string like `only Power VM system`

nasastry replied:
done

PraveenPenguin commented on perf/perf_24x7_all_events.py:
```diff
@@ -0,0 +1,135 @@
+            self.list_of_hv_24x7_events.append(line)
+
+        # Clear the dmesg, by that we can capture the delta at the end of the test.
+        process.run("dmesg -c")
```
`sudo=True ` is needed as dmesg only can be perform with root privileged  user

nasastry replied:
done

PraveenPenguin commented on perf/perf_24x7_all_events.py:
```diff
@@ -0,0 +1,135 @@
+
+    def tearDown(self):
+        # Collect the dmesg
+        process.run("dmesg -T")
```
are we doing any verification after collecting dmesg log ?

nasastry replied:
Just logging. Very rarely errors comes from this, in that case just wanted to log with out missing.

PraveenPenguin replied:
ah ok sure

nasastry commented:
@PraveenPenguin Taken care of the above comments. Please review. Thanks!!
